src/Chainweb/Pact/Backend/ChainwebPactDb.hs

```diff
@@ -433,192 +395,163 @@ createBlockHistoryTable =
+            [SInt (fromIntegral bh)]
+      exec' db "DELETE FROM [SYS:Modules] WHERE blockheight >= ?"
+            [SInt (fromIntegral bh)]
+      exec' db "DELETE FROM [SYS:Namespaces] WHERE blockheight >= ?"
+            [SInt (fromIntegral bh)]
+      exec' db "DELETE FROM [SYS:Pacts] WHERE blockheight >= ?"
+            [SInt (fromIntegral bh)]
 
 deleteHistory :: BlockHandler SQLiteEnv TxId
 deleteHistory = do
-    (BlockVersion bh v) <- gets _bsBlockVersion
+    bh <- gets _bsBlockHeight
     callDb "Deleting from BlockHistory, VersionHistory. Also get txid." $ \db -> do
         exec' db "DELETE FROM BlockHistory WHERE blockheight >= ?"
               [SInt (fromIntegral bh)]
-        exec' db "DELETE FROM VersionHistory WHERE version >= ?;"
-              [SInt (fromIntegral v)]
-        qry db "SELECT endingtxid FROM BlockHistory WHERE blockheight = ?"
+        -- exec' db "DELETE FROM VersionHistory WHERE version >= ?;" [SInt (fromIntegral v)]
+        t <- qry db "SELECT endingtxid FROM BlockHistory WHERE blockheight = ?"
             [SInt (fromIntegral $ pred bh)] [RInt]
           >>= fmap convert . expectSingleRowCol "txid after delete history"
+        return t
   where
     convert (SInt thing) = fromIntegral thing
     convert _ = error "deleteHistory: Something went wrong!"
 
 initSchema :: BlockHandler SQLiteEnv ()
 initSchema = withSavepoint DbTransaction $ do
     createBlockHistoryTable
-    createVersionHistoryTable
-    createUserTablesTable
+    createTableCreationTable
+    createTableMutationTable
     create (toTableName KeySets)
     create (toTableName Modules)
     create (toTableName Namespaces)
     create (toTableName Pacts)
   where
-    toTableName :: Domain k v -> TableName
-    toTableName = TableName . asString
+    toTableName :: Domain k v -> Utf8
+    toTableName = domainTableName
     create name = do
-        log "DDL" $ "initSchema: "  ++ show name
-        callDb "initSchema" $ \db ->
+        log "DDL" $ "initSchema: "  ++ toS name
+        callDb "initSchema" $ \db -> do
+            let tablename = name
+                indexname = name <> "_bh"
             exec_ db $
-            "CREATE TABLE IF NOT EXISTS"
-            <> Utf8 (toS (flip T.snoc ']' $ T.cons '[' $ asString name))
-            <> " (rowkey TEXT\
-               \, blockheight UNSIGNED BIGINT\
-               \, version UNSIGNED BIGINT\
-               \, txid UNSIGNED BIGINT\
-               \, rowdata BLOB\
-               \, UNIQUE(blockheight, rowkey, txid));"
+              "CREATE TABLE IF NOT EXISTS ["
+              <> name
+              <> "] (rowkey TEXT\
+                 \, blockheight UNSIGNED BIGINT NOT NULL\
+                 \, txid UNSIGNED BIGINT NOT NULL\
+                 \, rowdata BLOB NOT NULL\
+                 \, UNIQUE(blockheight, rowkey, txid));"
+            exec_ db $ "CREATE INDEX [" <>  indexname <> "] ON [" <> tablename <> "](blockheight);"
```

Comment:
Same here re blockheight index. What you want instead here is an index on `(rowkey, txid)`. Also, isn't this the same table as above in `createUserTable`? Refactor re "don't-repeat-yourself"

Reply:
Thanks for this suggestion. The code cleans up!